cpp/include/raft/spatial/knn/detail/ivf_pq_ann.cuh

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
```

Comment:
Why is this file named `ivf_pq_ann.cuh`? Wouldn't something like `ann.cuh` or `ann_implem.cuh` be better description since it implements `IVF`, `IVFPQ` and `IVFSQ`?

Reply:
Great point! I wanted to future-proof our API a little bit because we might be exposing other (potentially non-FAISS) implementations (such as graph, space partitioning or hash-based) in the future in addition to the quantization methods we're exposing here. Maybe we should rename the file to something like `ann_quantized.cuh`? I'm also inclined to suffix the file with `_faiss` (same for the brute force impl file) just so we can be more clear about the dependencies here. In the worst case, that should allow us to more easily separate out the header files in the future so those requiring the FAISS dependency become more clear to the callers. For now I think we can keep the header files the way they are since all of our dense nearest neighbors API requires FAISS. What do you think?

Reply:
Understood. Yes, both should do it. However, if we call this `ann_faiss.cuh`, other files would probably have to be called `ann_<lib>.cuh` which might be a bit confusing as to what the specific methods in these files are. That's why it may be preferable to call it `ann_<method>.cuh`. But again both are fine to me.

Reply:
My preference here would be to call it `ann_quantized_faiss.cuh`, actually, so it sounds like we're on the same page.